nemo/collections/asr/parts/submodules/cuda_graph_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,431 @@
+            cudart.cudaStreamUpdateCaptureDependenciesFlags.cudaStreamSetCaptureDependencies,
+        )
+    )
+    body_stream = torch.cuda.Stream()
+    previous_stream = torch.cuda.current_stream()
+    cu_call(
+        cudart.cudaStreamBeginCaptureToGraph(
+            body_stream.cuda_stream,
+            body_graph,
+            None,
+            None,
+            0,
+            cudart.cudaStreamCaptureMode.cudaStreamCaptureModeThreadLocal,
+        )
+    )
+    torch.cuda.set_stream(body_stream)
+
+    yield body_stream, body_graph
+
+    cuda.cuLaunchKernel(
+        while_loop_kernel, 1, 1, 1, 1, 1, 1, 0, body_stream.cuda_stream, while_loop_args.ctypes.data, 0
+    )
+
+    cudart.cudaStreamEndCapture(body_stream.cuda_stream)
+
+    torch.cuda.set_stream(previous_stream)
+
+
+class RNNTGreedyDecodeCudaGraph:
+    def __init__(self, max_symbols: int, cuda_device, caller):
+        if HAVE_CUDA_PYTHON:
+            check_cuda_python_cuda_graphs_conditional_nodes_supported()
+        else:
+            raise ValueError("Cannot instantiate RNNTGreedyDecodeCudaGraph without `pip install cuda-python`")
+
+        assert max_symbols is not None
+
+        self.symbols_added_t = torch.tensor(0, dtype=torch.int64, device=cuda_device)
+        self.max_symbols_t = torch.tensor(max_symbols, dtype=torch.int64, device=cuda_device)
+        self.max_symbols = max_symbols
+        self.not_all_blank_t = torch.tensor(True, dtype=torch.bool, device=cuda_device)
+
+        self.cuda_device = cuda_device
+
+        self.time_idx_t = torch.tensor(0, dtype=torch.int64, device=self.cuda_device)
+        self.max_out_len_t = torch.tensor(0, dtype=torch.int64, device=self.cuda_device)
+
+        self.encoder_output = None
+        self.encoder_output_length = None
+        self.f = None
+
+        # Reasonable default maximum time. 375 frames * (80ms / frame) = 30 seconds
+        # 80ms is the frame size of recent fastconformer models
+        # This does not affect correctness.
+        self.max_time = 375
+        self.batch_size = 0
+
+        self.scores_cpu = None
+        self.labels_cpu = None
+        self.graph = None
+        self.graph_exec = None
+
+        self.first_call = True
+
+        self.caller = caller
+
+    def _reinitialize(self, max_time, batch_size, encoder_output, encoder_output_length):
+        if self.first_call:
+            # We need to call the original _greedy_decode_blank_as_pad
+            # implementation at least once beforehand in order to make
+            # sure that pytorch is "initialized". Pytorch may be
+            # uninitialized if this code runs before any other pytorch
+            # operation in this process. Pytorch often lazily
+            # initializes things like a cudnnHandle_t via
+            # cudnnCreate(), which can involve synchronizing with the
+            # host. Such actions are not stream capturable to a graph.
+            self.caller._greedy_decode_blank_as_pad_loop_frames(
```

Comment:
Can we just return the results from this greedy decoding call (just check `self.first_call` in `__call__`), and run `_reinitialize` only on the second call?

Reply:
I suppose so... but does it really matter?